File: CSharpFundamentals/CSharpOOPbasics/EncapsulationEx/ShoppingSpree/Person.cs
using System;
using System.Collections.Generic;

namespace ShoppingSpree
{
    public class Person
    {
        private string name;
        private decimal money;

        public List<Product> Products { get; set; }

        public Person(string name, decimal money)
        {
            this.Name = name;
            this.Money = money;
            this.Products = new List<Product>();
        }

        public decimal Money
        {
            get { return money; }
            private set
            {
                if (value < 0)
                {
                    throw new ArgumentException("Money cannot be negative");
                }
                money = value;
            }
        }

        public string Name
        {
            get { return name; }
            private set
            {
                if (string.IsNullOrWhiteSpace(value))
                {
                    throw new ArgumentException("Name cannot be empty");
                }
                name = value;
            }
        }

        public void PurchaseProduct(Product product)
        {
            if (this.Money >= product.Cost)
            {
                Console.WriteLine($"{this.Name} bought {product.Name}");
                this.money -= product.Cost;
                Products.Add(product);
            }
            else
            {
                Console.WriteLine($"{this.Name} can't afford {product.Name}");
            }
        }

        public override string ToString()
        {
            if (this.Products.Count == 0)
            {
                return $"{this.Name} - Nothing bought";
            }
            else
            {
                string[] names = new string[this.Products.Count];
                for (int i = 0; i < Products.Count; i++)
                {
                    names[i] = Products[i].Name;
                }
                return $"{this.Name} - {string.Join(", ", names)}";
            }
        }
    }
}

File: JavaFundamentals/JavaOOP/Encapsulation-Ex/src/shopping_spree/Person.java
package shopping_spree;

import java.util.ArrayList;
import java.util.List;

public class Person {
    private String name;
    private double money;
    private List<Product> products;

    public Person(String name, double money) {
        this.setName(name);
        this.setMoney(money);
        this.products = new ArrayList<>();
    }

    private void setName(String name) {
        if (name.length() == 0 || name.trim().isEmpty()) {
            throw new IllegalArgumentException("Name cannot be empty");
        }
        this.name = name;
    }

    private void setMoney(double money) {
        if (money < 0) {
            throw new IllegalArgumentException("Money cannot be negative");
        }
        this.money = money;
    }

    public String getName() {
        return this.name;
    }

    public void buyProduct(Product product) {
        if (product.getCost() > this.money) {
            System.out.printf("%s can't afford %s%n",
                    this.getName(),
                    product.getName());
        } else {
            this.products.add(product);
            this.setMoney(this.money - product.getCost());
            System.out.printf("%s bought %s%n", this.getName(), product.getName());
        }
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        sb.append(this.getName()).append(" - ");
        if (this.products.size() == 0) {
            sb.append("Nothing bought");
        } else {
            for (int i = 0; i < this.products.size() - 1; i++) {
                sb.append(products.get(i).getName()).append(", ");
            }
            sb.append(products.get(products.size() - 1).getName());
        }
        return sb.toString().trim();
    }
}
